src/terminal/parser/OutputStateMachineEngine.cpp

```diff
@@ -1630,31 +1500,65 @@ bool OutputStateMachineEngine::_ParseHyperlink(const std::wstring_view string,
 
 // Routine Description:
 // - OSC 10, 11, 12 ; spec ST
-//      spec: a color in the following format:
-//          "rgb:<red>/<green>/<blue>"
-//          where <color> is two hex digits
+//      spec: The colors are specified by name or RGB specification as per XParseColor
+//
+//   It's possible to have multiple "spec", which by design equals to a series of OSC command
+//   with accumulated Ps. For example "OSC 10;color1;color2" is effectively an "OSC 10;color1"
+//   and an "OSC 11;color2".
+//
+//   However, we do not support the chaining of OSC 10-17 yet. Right now only the first parameter
+//   will take effect.
 // Arguments:
 // - string - the Osc String to parse
-// - rgb - receives the color that we parsed in the format: 0x00BBGGRR
+// - rgbs - receives the colors that we parsed in the format: 0x00BBGGRR
 // Return Value:
-// - True if a table index and color was parsed successfully. False otherwise.
+// - True if the first table index and color was parsed successfully. False otherwise.
 bool OutputStateMachineEngine::_GetOscSetColor(const std::wstring_view string,
-                                               DWORD& rgb) const noexcept
+                                               std::vector<DWORD>& rgbs) const noexcept
+try
 {
-    rgb = 0;
+    rgbs.clear();
 
     bool success = false;
 
-    DWORD color = 0;
-    success = s_ParseColorSpec(string, color);
+    size_t current = 0;
+    size_t expectedColorCount = 0;
 
-    if (success)
+    while (current < string.size())
     {
-        rgb = color;
+        std::optional<til::color> colorOptional;
+        const auto nextDelimiter = string.find(L";", current);
+        if (nextDelimiter == std::wstring::npos)
+        {
+            colorOptional = Utils::ColorFromXTermColor(string.substr(current));
+            // No more delimiter means no more left for parsing. This will effectively break out the loop.
+            current = string.size();
+        }
+        else
+        {
+            // Substr until the delimiter.
+            const auto length = nextDelimiter - current;
+            colorOptional = Utils::ColorFromXTermColor(string.substr(current, length));
+            // Skip this color and the delimiter. Start the next one.
+            current += length + 1;
+        }
+
+        expectedColorCount++;
+
+        if (colorOptional.has_value())
+        {
+            rgbs.push_back(colorOptional.value());
+            // Only mark the parsing as a success iff the first color is a success.
+            if (expectedColorCount == 1)
+            {
+                success = true;
+            }
+        }
     }
 
     return success;

```

Comment:
is this different from `return rgbs.size() > 0`?

Reply:
Actually there is. Say we have “10;invalid;valid”, the size of rgbs is 1. But right now we can only use the first color to set the OSC 10. I have no choice but to mark the return value as a failure. If one day we want to support separate handling of multi-param, then it should return true, incidating one of the colors is valid and should be act upon.